### DISCUSSION TITLE
`diff` a file that has been renamed?

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#4273] [Q&A] [answered]
How can I make `jj diff` show just the content changes in a file that has been renamed? `jj diff <newname>` shows that the entire contents are new.

**Comment by ilyagr** [comment]
Currently, the easiest way I can think of is `cp <newname> <oldname>; jj diff <oldname>`.

We (@fowles especially) are working on copy tracking, see e.g. https://github.com/martinvonz/jj/issues/3386. It's possible that some version of it will be usable soon, diff specifically is probably one of the first copy tracking features that's going to happen.

**Update:** Just to be clear, there are no promises. I pinged Matt in the sense of "hey, your work is appreciated", not in the sense of "you are now responsible for doing ABC on XYZ schedule".

**Reply by fowles** [reply, marked as answer]
the last PR in the series is https://github.com/martinvonz/jj/pull/4279 after that, it is just waiting for a new release.  So I feel fairly confident in saying it will be ready in time for the next release.

If you use an external diff tool, it must be run in file-by-file mode.

**Reply by BatmanAoD** [reply]
Now that copy-tracking has been released (🎉), is there a way to configure conflict-resolution such that, when a rename happens on one or both sides of a diff, the last version of the pre-rename file is used as the `$base`?

I have a branch where I renamed a file, and that file changed in `trunk()` while I was working on the branch. I thought I could potentially make the merge easier by making a separate MR that renamed the file to match the name in my branch (so this was the "both sides of a diff" case). But conflict resolution, regardless of whether I used `new` or `rebase`, treated both sides as fully unique, with no shared content in the "base" revision.

What I ended up doing was:

* I duplicated the `trunk()` commit that my branch was based on (i.e. the shared ancestor commit, in which the file had the old name).
* I edited the duplicate commit by *copying* the file to the new name.
* I rebased my branch onto the newly edited commit, then resolved the resulting conflict in favor of my branch (i.e. I accepted all changes from my branch and ignored the "left" side of the conflict).
* I then rebased my branch again (excluding the duplicate commit), this time onto `trunk()`. This gave ma a proper diff to resolve.

**Reply by yuja** [reply]
> is there a way to configure conflict-resolution such that, when a rename happens on one or both sides of a diff, the last version of the pre-rename file is used as the $base?

No, rename tracking isn't add to merge/rebase logic yet.
